src/supervisor/mod.rs

```diff
@@ -198,135 +211,320 @@ async fn terminate_children<'a>(
         self,
         parent_ctx: &context::Context,
-        parent_name: &'a str,
     ) -> Result<Supervisor, (Spec, Arc<StartError>)> {
-        let (ctx, _terminate_supervisor) = Context::with_cancel(parent_ctx);
+        // start can only be called from a root supervisor, ergo, parent_name is blank
+        let parent_name = "".to_owned();
 
-        let mut ev_notifier = self.meta.ev_notifier.clone();
-        let meta = self.meta;
+        // build sup_runtime_name for tracing purposes
+        let sup_runtime_name = format!("{}/{}", parent_name, self.meta.name);
 
-        let sup_runtime_name = format!("{}/{}", parent_name, meta.name);
-        let result = start_children(
-            &ctx,
-            meta.ev_notifier.clone(),
-            &sup_runtime_name,
-            meta.order,
-            self.children,
-        )
-        .await;
-
-        // TODO: monitoring should start here
-        match result {
-            Err((children_specs, err)) => {
-                let err = Arc::new(err);
-                ev_notifier
-                    .supervisor_start_failed(&sup_runtime_name, err.clone())
-                    .await;
-                let spec = Spec {
-                    meta,
-                    children: children_specs,
-                };
-                return Err((spec, err));
-            }
-            Ok(children) => {
-                ev_notifier.supervisor_started(&sup_runtime_name).await;
+        // cancel_fn is used when Supervisor#terminate() is called
+        let (ctx, termination_handle) = parent_ctx.with_cancel();
+
+        // // minimum buffer size is 10, otherwise use the number of children
+        // let children_count = std::cmp::min(10, self.children.len());
+
+        // TODO
+        // // notify is used to listen to errors on child nodes
+        // let (notify_send, notify_recv) = broadcast::channel(children_count);
+
+        // TODO
+        // // ctrl is used to keep track of requests from client APIs (e.g. spawn child)
+        // let (ctrl_send, ctrl_recv) = broadcast::channel(children_count);
+
+        // started is used when waiting for the worker to signal it has started
+        let (started_send, started_recv) = oneshot::channel();
+        let start_notifier = StartNotifier::from_oneshot(started_send);
+
+        let spec_children = self.children;
+
+        // clone all variables that we are going to move to the monitor loop routine
+        let meta = self.meta.clone();
+        let sup_runtime_name1 = sup_runtime_name.clone();
+
+        let join_handle = task::spawn(run_supervisor_monitor(
+            ctx,
+            meta,
+            spec_children,
+            Some(start_notifier),
+            sup_runtime_name1,
+        ));
+
+        // We wait for the supervisor to get started, and from there we return a
+        // Supervisor record to the client, which can terminate the supervisor,
+        // or wait for it to finish
+        return match started_recv.await {
```

Comment:
explicit `return` here is unnecessary. Just remove the ; at the end of the match block as well.